### DISCUSSION TITLE
Flushing during Session.execute

### DISCUSSION THREAD

**Discussion by rbhanot4739** [sqlalchemy/sqlalchemy#11343] [Usage Questions]
Hello Community,

I was going through the documentation on [Mapper level flush events](https://docs.sqlalchemy.org/en/20/orm/session_events.html#mapper-level-flush-events) and there is this note mentioned 

> It is important to note that these events apply only to the [session flush operation](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-flushing) , and not to the ORM-level INSERT/UPDATE/DELETE functionality described at [ORM-Enabled INSERT, UPDATE, and DELETE statements](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-expression-update-delete). To intercept ORM-level DML, use the [SessionEvents.do_orm_execute()](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) event.

So when I started reading the [Flushing](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#flushing) section it mentions that flushing also happens during `Session.execute()`.

This got me confused since all of the ORM enabled statements make use of `Session.execute()` or similar like `Session.Scalars()` which would implicitly mean that Flushing should happen on ORM enabled statements as well ?

Infact docs for [do_orm_execute](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) also has this note which says 

> Also, this event hook does not apply to queries that are emitted internally within the ORM flush process, i.e. the process described at [Flushing](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-flushing); to intercept steps within the flush process, see the event hooks described at [Persistence Events](https://docs.sqlalchemy.org/en/20/orm/session_events.html#session-persistence-events) as well as [Mapper-level Flush Events](https://docs.sqlalchemy.org/en/20/orm/session_events.html#session-persistence-mapper).

And since Flushing anyways happens at `Session.execute()` as mentioned above, I am little confused and feeling these two pieces of documentation are contradicting each other. 

Can someone please give some more clarity on this ?

**Comment by zzzeek** [comment]
> 
> > It is important to note that these events apply only to the [session flush operation](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-flushing) , and not to the ORM-level INSERT/UPDATE/DELETE functionality described at [ORM-Enabled INSERT, UPDATE, and DELETE statements](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-expression-update-delete). To intercept ORM-level DML, use the [SessionEvents.do_orm_execute()](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.do_orm_execute) event.
> 
> So when I started reading the [Flushing](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#flushing) section it mentions that flushing also happens during `Session.execute()`.
> 
> This got me confused since all of the ORM enabled statements make use of `Session.execute()` or similar like `Session.Scalars()` which would implicitly mean that Flushing should happen on ORM enabled statements as well ?

it's autoflush.   it occurs **first**, before the main SQL statement you want to run happens.   within the flush, it potentially runs other SQL statements but these statements do not pass through the do_orm_execute() process.


> 
> > Also, this event hook does not apply to queries that are emitted internally within the ORM flush process, i.e. the process described at [Flushing](https://docs.sqlalchemy.org/en/20/orm/session_basics.html#session-flushing); to intercept steps within the flush process, see the event hooks described at [Persistence Events](https://docs.sqlalchemy.org/en/20/orm/session_events.html#session-persistence-events) as well as [Mapper-level Flush Events](https://docs.sqlalchemy.org/en/20/orm/session_events.html#session-persistence-mapper).
> 
> And since Flushing anyways happens at `Session.execute()` as mentioned above, I am little confused and feeling these two pieces of documentation are contradicting each other.

the flush happens and its INSERT/UPDATE/DELETE statements, if any, will occur, but they are not individually intercepted by the do_orm_execute() event hook.   if the flush emits an INSERT for a table, and you have a do_orm_execute() hook, you will not receive any message for that specific statement in the hook.

**Comment by rbhanot4739** [comment]
Thank you @zzzeek  for the answer, however I have few followup questions on that. 

Scenario 1
- Create a session and bunch of model objects.
- Add these objects to session using `session.add_all()`.
- Commit these objects to db with `session.commit()`

So there is an implicit flush happening as part of `commit()`.

Secario 2
 - Use the ORM enabled Insert statement with `session.execute()` to add the objects to the db.

As you mentioned above there is an autoflush happening in this case as well. So what exactly is the difference b/w the flush that happens as part of `commit` and the autoflush that happens with `execute` ?

So  does that mean `do_orm_execute` works with scenario 2 and mapper level flush events works with scenarion 1 ?

Also I assume the underlyiing sql emitted in both the cases would be same right ? If so how does the instrumentation detect that ORM level INSERT/UPDTE/DELETE are not be passed to the mapper level flush events and queries that are emitted internally within the ORM flush process are not to be passed to `do_orm_execute` ?

**Reply by zzzeek** [reply]
flush() will emit DML statements for mapped objects that are part of the session using `session.add()`.    an ORM enabled INSERT statement OTOH doesn't have any mapped objects in play when it runs, so there is nothing within that statement that the flush() process has any interaction with.  the flush runs here for any *other* objects that might be pending before the statement actually runs.

> So does that mean do_orm_execute works with scenario 2 and mapper level flush events works with scenarion 1 ?

pretty much

> Also I assume the underlyiing sql emitted in both the cases would be same right ?

well in the sense of INSERT statements with particular values sure, but as far as if the statements are batched and what columns are used in RETURNING and such, probably not.   turn on echo=True on your engine and just observe the differences.


>  If so how does the instrumentation detect that ORM level INSERT/UPDTE/DELETE are not be passed to the mapper level flush events and queries that are emitted internally within the ORM flush process are not to be passed to do_orm_execute ?

when the flush process emits SQL it isn't using `session.execute()`, it's using the `.execute()` method of a `Connection` object which is lower level than the `Session`.
